acceptance/tests/pxp-module-puppet/run_puppet_result_failed.rb

```diff
@@ -0,0 +1,53 @@
+require 'pxp-agent/test_helper.rb'
+require 'puppet/acceptance/environment_utils'
+
+test_name 'C93063 - pxp-module-puppet run expect puppet failure' do
+  extend Puppet::Acceptance::EnvironmentUtils
+
+  app_type = File.basename(__FILE__, '.*')
+  tmp_environment   = mk_tmp_environment(app_type)
+
+  step 'On master, create a new environment that will result in Failed' do
+    site_manifest = File.join(environmentpath, tmp_environment, 'manifests', 'site.pp')
+    create_remote_file(master, site_manifest, <<-SITEPP)
+node default {
+  fail('intended failure')
+}
+SITEPP
+  end
+
+  step 'Ensure each agent host has pxp-agent running and associated' do
+    agents.each do |agent|
+      on agent, puppet('resource service pxp-agent ensure=stopped')
+      create_remote_file(agent, pxp_agent_config_file(agent), pxp_config_json_using_puppet_certs(master, agent).to_s)
+      on agent, puppet('resource service pxp-agent ensure=running')
+      assert(is_associated?(master, "pcp://#{agent}/agent"),
+             "Agent #{agent} with PCP identity pcp://#{agent}/agent should be associated with pcp-broker")
+    end
+  end
+
+  step "Send an rpc_blocking_request to all agents" do
+    target_identities = []
+    agents.each do |agent|
+      target_identities << "pcp://#{agent}/agent"
+    end
+    responses = nil # Declare here so not local to begin/rescue below
+    begin
+      responses = rpc_blocking_request(master, target_identities,
+                                      'pxp-module-puppet', 'run',
+                                      {:env => [], :flags => ['--environment', tmp_environment,
+                                                              '--onetime']
+                                      })
+    rescue => exception
+      fail("Exception occurred when trying to run Puppet on all agents: #{exception.message}")
+    end
+    agents.each do |agent|
+      step "Check Run Puppet response for #{agent}" do
+        identity = "pcp://#{agent}/agent"
+        action_result = responses[identity][:data]["results"]
+        assert(action_result.has_key?('status'), "Results for pxp-module-puppet run on #{agent} should contain a 'status' field")
+        assert_equal('failed', action_result['status'], "Result of pxp-module-puppet run on #{agent} should be 'failed'. was: '#{action_result["status"]}'")
```

Comment:
Using the fail() function is a little different to what this test case intended - it causes a parsing error creating the catalog, and the agent errors out because it did not get a catalog to run with....

`[root@wlu11s2tuopmhnu ~]# puppet agent -t --environment failure
Info: Using configured environment 'failure'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Evaluation Error: Error while evaluating a Function Call, Intended failure at /etc/puppetlabs/code/environments/failure/manifests/site.pp:2:3 on node wlu11s2tuopmhnu.delivery.puppetlabs.net
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
[root@wlu11s2tuopmhnu ~]# echo $?
1`

This is different to a failure occurring during the catalog run e.g. if I try to ensure a file inside a directory that doesn't exist:

`[root@wlu11s2tuopmhnu ~]# puppet agent -t --environment failure
Info: Using configured environment 'failure'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Caching catalog for wlu11s2tuopmhnu.delivery.puppetlabs.net
Info: Applying configuration version '1461933225'
Error: Could not set 'file' on ensure: No such file or directory @ rb_sysopen - /tmp/badfolder/jimmy at 2:/etc/puppetlabs/code/environments/failure/manifests/site.pp
Error: Could not set 'file' on ensure: No such file or directory @ rb_sysopen - /tmp/badfolder/jimmy at 2:/etc/puppetlabs/code/environments/failure/manifests/site.pp
Wrapped exception:
No such file or directory @ rb_sysopen - /tmp/badfolder/jimmy
Error: /Stage[main]/Main/Node[default]/File[/tmp/badfolder/jimmy]/ensure: change from absent to file failed: Could not set 'file' on ensure: No such file or directory @ rb_sysopen - /tmp/badfolder/jimmy at 2:/etc/puppetlabs/code/environments/failure/manifests/site.pp
Notice: Applied catalog in 0.01 seconds
[root@wlu11s2tuopmhnu ~]# echo $?
4`

I think there are two potential test cases here; using fail() shows what PXP does if the catalog compilation fails; the latter case shows what PXP does if a failure occurs during catalog execution


btw: +1 on not inspecting logs unless absolutely necessary